ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,938 @@ PHP_FUNCTION(bcscale)
 /* }}} */
 
 
+
+static zend_class_entry *bcmath_number_ce;
+static zend_object_handlers bcmath_number_obj_handlers;
+
+static zend_always_inline bcmath_number_obj_t *get_bcmath_number_from_obj(const zend_object *obj)
+{
+	return (bcmath_number_obj_t*)((char*)(obj) - XtOffsetOf(bcmath_number_obj_t, std));
+}
+
+static zend_always_inline bcmath_number_obj_t *get_bcmath_number_from_zval(const zval *zv)
+{
+	return get_bcmath_number_from_obj(Z_OBJ_P(zv));
+}
+
+static zend_always_inline zend_string *bcmath_number_value_to_str(bcmath_number_obj_t *intern)
+{
+	if (intern->value == NULL) {
+		intern->value = bc_num2str_ex(intern->num, intern->scale);
+	}
+	return intern->value;
+}
+
+static zend_object *bcmath_number_create(zend_class_entry *ce)
+{
+	bcmath_number_obj_t *intern = zend_object_alloc(sizeof(bcmath_number_obj_t), ce);
+
+	zend_object_std_init(&intern->std, ce);
+	object_properties_init(&intern->std, ce);
+
+	intern->num = NULL;
+	intern->value = NULL;
+	intern->scale = 1;
+
+	return &intern->std;
+}
+
+static void bcmath_number_free(zend_object *obj)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	if (intern->num) {
+		bc_free_num(&intern->num);
+		intern->num = NULL;
+	}
+	if (intern->value) {
+		zend_string_release(intern->value);
+		intern->value = NULL;
+	}
+	zend_object_std_dtor(&intern->std);
+}
+
+static zend_object *bcmath_number_clone(zend_object *obj)
+{
+	bcmath_number_obj_t *original = get_bcmath_number_from_obj(obj);
+	bcmath_number_obj_t *clone = get_bcmath_number_from_obj(bcmath_number_create(bcmath_number_ce));
+
+	clone->num = bc_copy_num(original->num);
+	if (original->value) {
+		clone->value = zend_string_copy(original->value);
+	}
+	clone->scale = original->scale;
+
+	return &clone->std;
+}
+
+static HashTable *bcmath_number_get_properties_for(zend_object *obj, zend_prop_purpose purpose)
+{
+	switch (purpose) {
+		case ZEND_PROP_PURPOSE_DEBUG:
+		case ZEND_PROP_PURPOSE_SERIALIZE:
+		case ZEND_PROP_PURPOSE_VAR_EXPORT:
+		case ZEND_PROP_PURPOSE_JSON:
+		case ZEND_PROP_PURPOSE_ARRAY_CAST:
+			break;
+		default:
+			return zend_std_get_properties_for(obj, purpose);
+	}
+
+	zval zv;
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	HashTable *props = zend_array_dup(zend_std_get_properties(obj));
+
+	ZVAL_STR_COPY(&zv, bcmath_number_value_to_str(intern));
+	zend_hash_str_update(props, "value", sizeof("value")-1, &zv);
+	ZVAL_LONG(&zv, intern->scale);
+	zend_hash_str_update(props, "scale", sizeof("scale")-1, &zv);
+
+	return props;
+}
+
+static zval *bcmath_number_write_property(zend_object *obj, zend_string *name, zval *value, void **cache_slot)
+{
+	if (zend_string_equals_literal(name, "value") || zend_string_equals_literal(name, "scale")) {
+		zend_throw_error(NULL, "Cannot modify readonly property %s::$%s", ZSTR_VAL(obj->ce->name), ZSTR_VAL(name));
```

Comment:
There's a standard method for this error message.
Besides that, do you even need this write_property handler if you use `@strict-properties` ?

Reply:
It doesn't actually use the property, it acts like a virtual property, so if use the standard error messages will get error messages about initialization.

Reply:
Ok I see. Please use `zend_readonly_property_modification_error_ex` for the readonly message though. I don't think we have one for the dynamic properties one though but I still think you can fall back to the default write function to trigger that.

Reply:
Fixed in https://github.com/php/php-src/pull/13741/commits/56f5b82bc72f0e492a141ec230f40af8f8cdb7a1